src/hooks/useEditorView.ts

```diff
@@ -59,70 +36,86 @@ function withFlushedDispatch(
+  if (process.env.NODE_ENV !== "production") {
+    if (
+      props.defaultState !== undefined &&
+      props.state !== undefined &&
+      !didWarnValueDefaultValue
+    ) {
+      console.error(
+        "A component contains a ProseMirror editor with both value and defaultValue props. " +
+          "ProseMirror editors must be either controlled or uncontrolled " +
+          "(specify either the state prop, or the defaultState prop, but not both). " +
+          "Decide between using a controlled or uncontrolled ProseMirror editor " +
+          "and remove one of these props. More info: " +
+          "https://reactjs.org/link/controlled-components"
+      );
+      didWarnValueDefaultValue = true;
+    }
+  }
 
-  const stateProp = "state" in editorProps ? editorProps.state : undefined;
+  const defaultState = props.defaultState ?? EMPTY_STATE;
+  const [_state, setState] = useState<EditorState>(defaultState);
+  const state = props.state ?? _state;
 
-  const state =
-    "defaultState" in editorProps
-      ? editorProps.defaultState
-      : editorProps.state;
+  const {
+    componentEventListenersPlugin,
+    registerEventListener,
+    unregisterEventListener,
+  } = useComponentEventListeners();
 
-  const nonStateProps = Object.fromEntries(
-    Object.entries(editorProps).filter(
-      ([propName]) => propName !== "state" && propName !== "defaultState"
-    )
+  const plugins = useMemo(
+    () => [...(props.plugins ?? []), componentEventListenersPlugin],
+    [props.plugins, componentEventListenersPlugin]
   );
 
-  useLayoutEffect(() => {
-    return () => {
-      if (view) {
-        view.destroy();
+  function dispatchTransaction(this: EditorView, tr: Transaction) {
+    flushSync(() => {
+      if (props.dispatchTransaction) {
+        props.dispatchTransaction.call(this, tr);
+      } else {
+        setState((s) => s.apply(tr));
       }
-    };
-  }, [view]);
-
-  useLayoutEffect(() => {
-    if (view && view.dom !== mount) {
-      setView(null);
-      return;
-    }
-
-    if (!mount) {
-      return;
-    }
+    });
+  }
 
-    if (!view) {
-      setView(
-        new EditorView(
-          { mount },
-          {
-            ...editorProps,
-            state,
-          }
-        )
-      );
-      return;
-    }
-  }, [editorProps, mount, state, view]);
+  const directEditorProps = {
+    ...props,
+    state,
+    plugins,
+    dispatchTransaction,
+  };
 
-  useLayoutEffect(() => {
-    view?.setProps(nonStateProps);
-  }, [view, nonStateProps]);
+  const [view, setView] = useState<EditorView | null>(null);
 
+  // This effect runs on every render and handles the view lifecycle.
+  // eslint-disable-next-line react-hooks/exhaustive-deps
   useLayoutEffect(() => {
-    if (stateProp) view?.setProps({ state: stateProp });
-  }, [view, stateProp]);
-
-  return view;
+    if (view) {
+      if (view.dom === mount) {
+        view.setProps(directEditorProps);
+      } else {
+        view.destroy();
```

Comment:
In the previous implementation, we destroy the view in the `useLayoutEffect` cleanup function. Trying to understand this: this implementation is functionally equivalent, because if the `mount` effect argument changes, this should have the same effect of unmounting (and subsequently, remounting with the logic below) by calling `view.destroy()` at the right moment?
It might be nice to have a test for unmounting.

Reply:
Yeah, good call out.

Setting the state within an effect triggers an immediate re-render, and since this is happening at an ancestor of any component that could access the view it should be the case that they've all run their effects already. So, the net result either way is an immediate re-render of the descendants with a null view.

I guess this way we'll destroy the node views first, which could cause portals to unregister, thereby forcing an update of an even higher component first.

I think it's worth backing this change out of this PR and considering this part more carefully. Thanks.

Reply:
I've backed that out.